### PR TITLE
fix(migration): migrate `deprecated_tags` to new tagging model

### PR DESCRIPTION
**Pull request by alexkim205**
## Changes

Followup AI from https://github.com/PostHog/postmortem/pull/13

This PR aims to migrate old production tags that are stuck in `deprecated_tags` columns to the new TaggedItem-Tag system.

You'll probably want to give this ticket a 👀 to get the full context for why this migration came out the way it did. https://github.com/PostHog/posthog/issues/8771.

~24.5K of the diff is made up of snapshots.

*Stay up-to-date with our [coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

Wrote a migration test that covers these edge cases:
- Already existing tag
- Already existing taggeditem
- Duplicate tags within object
- Duplicate tags across objects within migration
- Tags from different teams
- Number of tags > `batch_size`
- Nullish tags

### CODE REVIEW

File: ee/test/test_migration_0012.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            id=self.property_definition_without_tags.id
+        )
+        self.assertEqual(property_definition_without_tags.tagged_items.count(), 0)
+
+        # By the end of the migration, the total count for team 2 should be
+        # Tags = team2_total_property_definitions + 2 + team1_tags
+        # TaggedItems = team2_total_property_definitions * 3 + team1_taggeditems
+        self.assertEqual(Tag.objects.filter(team_id=self.team.id).all().count(), 5)
+        self.assertEqual(Tag.objects.all().count(), self.team2_total_property_definitions + 2 + 5)
+        self.assertEqual(TaggedItem.objects.all().count(), self.team2_total_property_definitions * 3 + 7)
```
Comment (alexkim205) [inline ee/test/test_migration_0012.py:84]: It would be the icing on the cake if we could also assert on the number of transactions that have occurred, but adding boilerplate to support that seems like scope creep. Any suggestions on how to do it easily are welcome!

**Review by macobo — COMMENTED** [review]
Only reviewing this from a code/migration issues standpoint, didn't take a look at correctness.

There's nothing immediately suspect here besides the try-catch.

I suggest running the loops (without actually creating anything) once against production as well as that will probably reveal issues. https://posthog.com/handbook/engineering/production-access

File: ee/migrations/0012_migrate_tags_v2.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    TaggedItem = apps.get_model("posthog", "TaggedItem")
+    EnterpriseEventDefinition = apps.get_model("ee", "EnterpriseEventDefinition")
+    EnterprisePropertyDefinition = apps.get_model("ee", "EnterprisePropertyDefinition")
+
+    createables: List[Tuple[Any, Any]] = []
+    batch_size = 1000
+
+    try:
+        # Collect insight tags and taggeditems
+        for [tags, team_id, id] in EnterpriseEventDefinition.objects.exclude(
```
Comment (macobo) [inline ee/migrations/0012_migrate_tags_v2.py:23]: Nit: This is unpythonic. Either do `(tags, team_id, id)` or `tags, team_id, id`

File: ee/migrations/0012_migrate_tags_v2.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            unique_tags = set([tagify(t) for t in tags if isinstance(t, str) and t.strip() != ""])
+            if unique_tags:
+                for tag in unique_tags:
+                    createables.append((Tag(name=tag, team_id=team_id), TaggedItem(event_definition_id=id)))
+
+        # Collect dashboard tags and taggeditems
+        for [tags, team_id, id] in EnterprisePropertyDefinition.objects.exclude(
+            Q(deprecated_tags__isnull=True) | Q(deprecated_tags=[]),
+        ).values_list("deprecated_tags", "team_id", "id"):
+            unique_tags = set([tagify(t) for t in tags if isinstance(t, str) and t.strip() != ""])
```
Comment (macobo) [inline ee/migrations/0012_migrate_tags_v2.py:35]: Nit: Remove `[`

File: posthog/migrations/0217_migrate_tags_v2.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            for pair in createables[i : (i + batch_size)]:
+                if pair[0].id in created_tags:
+                    pair[1].tag_id = created_tags[pair[0].id].id
+                else:
+                    pair[1].tag_id = Tag.objects.get(name=pair[0].name, team_id=pair[0].team_id).id
+
+        # Create tag <-> item relationships, ignoring conflicts
+        TaggedItem.objects.bulk_create([pair[1] for pair in createables], ignore_conflicts=True, batch_size=batch_size)
+
+    except Exception as e:
```
Comment (guidoiaquinti) [inline posthog/migrations/0217_migrate_tags_v2.py:64]: Ditto

File: ee/test/test_migration_0012.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
+from django.db import transaction
+from django.db.models import Q
+
+from posthog.models import Tag as TagModel
+from posthog.models import TaggedItem as TaggedItemModel
+from posthog.models import Team
+from posthog.test.base import TestMigrations
+
+
+class TagsTestCase(TestMigrations):
```
Comment (macobo) [inline ee/test/test_migration_0012.py:10]: If you're adding tests I suggest using the snapshot functions here as well to document the generated SQL

File: posthog/test/base.py [hunk 6]
```diff
@@ -1,4 +1,5 @@
 import inspect
+import logging
```
Comment (guidoiaquinti) [inline posthog/test/base.py:2]: Is this needed?

File: posthog/utils.py [hunk 7]
```diff
@@ -5,6 +5,7 @@
 import gzip
 import hashlib
 import json
+import logging
```
Comment (guidoiaquinti) [inline posthog/utils.py:8]: Can we use `structlog` as per [docs](https://posthog.com/docs/contribute/coding-conventions#format) ? 🙏

File: ee/migrations/0012_migrate_tags_v2.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        # scenario, all tags already exist and get is made for every tag.
+        for i in range(0, len(tags_to_create), batch_size):
+            batch = tags_to_create[i : (i + batch_size)]
+
+            # Find tags that were created, and not already existing
+            created_tags = Tag.objects.in_bulk([t.id for t in batch])
+
+            # Tags that are in `tags_to_create` but not in `created_tags` are tags that already exist
+            # in the db and must be fetched individually.
+            for pair in createables[i : (i + batch_size)]:
```
Comment (macobo) [inline ee/migrations/0012_migrate_tags_v2.py:55]: Nit: `for tag, tagged_item in `

File: ee/migrations/0012_migrate_tags_v2.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        ).values_list("deprecated_tags", "team_id", "id"):
+            unique_tags = set([tagify(t) for t in tags if isinstance(t, str) and t.strip() != ""])
+            if unique_tags:
+                for tag in unique_tags:
+                    createables.append((Tag(name=tag, team_id=team_id), TaggedItem(property_definition_id=id)))
+
+        # Attempts to create tags in bulk while ignoring conflicts. bulk_create does not return any data
+        # about which tags were ignored and created, so we must take care of this manually.
+        tags_to_create = [pair[0] for pair in createables]
+        Tag.objects.bulk_create(tags_to_create, ignore_conflicts=True, batch_size=batch_size)
```
Comment (hazzadous) [inline ee/migrations/0012_migrate_tags_v2.py:43]: Note that this will try to insert a tag for each tag item. I don't think it will be an issue but it's more than we need.

File: ee/migrations/0012_migrate_tags_v2.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+            Q(deprecated_tags__isnull=True) | Q(deprecated_tags=[]),
+        ).values_list("deprecated_tags", "team_id", "id"):
+            unique_tags = set([tagify(t) for t in tags if isinstance(t, str) and t.strip() != ""])
+            if unique_tags:
+                for tag in unique_tags:
+                    createables.append((Tag(name=tag, team_id=team_id), TaggedItem(property_definition_id=id)))
+
+        # Attempts to create tags in bulk while ignoring conflicts. bulk_create does not return any data
+        # about which tags were ignored and created, so we must take care of this manually.
+        tags_to_create = [pair[0] for pair in createables]
```
Comment (hazzadous) [inline ee/migrations/0012_migrate_tags_v2.py:42]: Rather than using just `pair` can we unpack into meaningfully named vars, and use `_` if we do not need to reference, e.g.

```
tags_to_create = [tag for (tag, _) in creatables]
```

File: ee/migrations/0012_migrate_tags_v2.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        Tag.objects.bulk_create(tags_to_create, ignore_conflicts=True, batch_size=batch_size)
+
+        # Associate tag ids with tagged_item objects in batches. Best case scenario all tags are new. Worst case
+        # scenario, all tags already exist and get is made for every tag.
+        for i in range(0, len(tags_to_create), batch_size):
+            batch = tags_to_create[i : (i + batch_size)]
+
+            # Find tags that were created, and not already existing
+            created_tags = Tag.objects.in_bulk([t.id for t in batch])
```
Comment (hazzadous) [inline ee/migrations/0012_migrate_tags_v2.py:51]: If bulk create doesn't return anything, won't this t.id always be `None`?

Also, I don't think we need two layers of batching here, but batch up the taggeditems, for each batch fetch the tag_id associated with (name, team_id), bulk insert that batch.

File: ee/test/test_migration_0012.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+                    name=f"batch_prop_{_tag}",
+                    team_id=self.team2.id,
+                    deprecated_tags=[_tag, "existing tag", "tag to create"],
+                )
+        TaggedItemModel.objects.create(
+            tag=tag2,
+            property_definition_id=EnterprisePropertyDefinition.objects.filter(team_id=self.team2.id).first().id,
+        )
+
+    def test_tags_migrated(self):
```
Comment (hazzadous) [inline ee/test/test_migration_0012.py:60]: Can we run the migration twice?

non-blocking: I'm not sure how that fits in with the `TestMigration` structure, but tbh I'd prefer to just perform the migrations explicitly in these tests rather implicitly somewhere in the class hierarchy

File: posthog/migrations/0217_migrate_tags_v2.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
+# Generated by Django 3.2.5 on 2022-03-01 23:41
```
Comment (hazzadous) [inline posthog/migrations/0217_migrate_tags_v2.py:1]: What's with the two migrations? Is this just because they are in diff django apps?

**Review by hazzadous — DISMISSED** [review]

File: posthog/utils.py (hunk 7 above)
Comment (alexkim205) [reply posthog/utils.py:8]: Ah thank you for the link!

File: ee/migrations/0012_migrate_tags_v2.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+            for pair in createables[i : (i + batch_size)]:
+                if pair[0].id in created_tags:
+                    pair[1].tag_id = created_tags[pair[0].id].id
+                else:
+                    pair[1].tag_id = Tag.objects.get(name=pair[0].name, team_id=pair[0].team_id).id
+
+        # Create tag <-> item relationships, ignoring conflicts
+        TaggedItem.objects.bulk_create([pair[1] for pair in createables], ignore_conflicts=True, batch_size=batch_size)
+
+    except Exception as e:
```
[the remark this answers is not in the document]
Comment (alexkim205) [reply ee/migrations/0012_migrate_tags_v2.py:64]: Yeah that makes a lot of sense. I think the issue with the first version of the migration was that Django wasn't logging any errors which made it difficult to debug what was going on, so I just threw in a try catch that would intercept the error, print it, and then throw it again (which I forgot to do 😅).

There isn't a specific error that I wanted to recover from here and adding logging info like you suggested below will give us some clue where an error happens so I think it makes sense to remove this try catch entirely.

File: posthog/migrations/0217_migrate_tags_v2.py (hunk 13 above)
Comment (alexkim205) [reply posthog/migrations/0217_migrate_tags_v2.py:1]: It's a weird consequence/code smell of splitting out event_definitions and property_definitions into a seperate enterprise app, which requires them to be migrated separately.

File: ee/migrations/0012_migrate_tags_v2.py (hunk 11 above)
Comment (alexkim205) [reply ee/migrations/0012_migrate_tags_v2.py:51]: > If bulk create doesn't return anything, won't this t.id always be None?

Good q, if `bulk_create` doesn't create any new tags, `in_bulk` will return an empty list which indicates that all the tags to be created already existed. I don't think `t` here will ever be an empty object since any object returned here must already exist in the table.

> Also, I don't think we need two layers of batching here, but batch up the taggeditems, for each batch fetch the tag_id associated with (name, team_id), bulk insert that batch.

👍 makes sense!

File: ee/migrations/0012_migrate_tags_v2.py (hunk 9 above)
Comment (alexkim205) [reply ee/migrations/0012_migrate_tags_v2.py:43]: It's suboptimal and I would've preferred to do an in_bulk check that can fetch all tags that already exist before bulk creating them like this. I couldn't think of a way to do this using django's limited API here, so just opted for the simplest solution here.

File: ee/test/test_migration_0012.py (hunk 12 above)
Comment (alexkim205) [reply ee/test/test_migration_0012.py:58]: I think this will require some custom boilerplate since `TestMigration` intercepts the one time `setUp` call to perform the migration. I agree it'd be nice to have separate setup, execute, and teardown methods for each test but I think it'll be fine for now.

**Comment by alexkim205** [comment]
Appreciate the reviews! Addressed all feedback.

### re: testing in prod

@macobo Unfortunately it looks like I can't run the read-only part of the migration on production data, since I'm unable to access the deprecated `deprecated_tags` field from the Django shell. The `django_deprecate_fields` library only allows accessing this field in migration related commands, which makes sense in practice but doesn't work for when we're trying to spot test in a deprecated field.

```python
// site-packages/django_deprecate_fields/deprecate_field.py
...
if not set(sys.argv) & {"makemigrations", "migrate", "showmigrations"}:
    return DeprecatedField(return_instead)
...
```

The next closest thing that came to mind was to run the SQL queries that are generated by the migration on Metabase. All queries didn't require any batching since the total counts in production were all well under the batch_size (1000).

```
  SELECT COUNT(*) AS "__count"
  FROM "posthog_dashboarditem"
  WHERE NOT (("posthog_dashboarditem"."deprecated_tags" IS NULL
              OR ("posthog_dashboarditem"."deprecated_tags" = ARRAY[]::varchar(32)[]
                  AND "posthog_dashboarditem"."deprecated_tags" IS NOT NULL)))

 > 113
```

```
  SELECT COUNT(*) AS "__count"
  FROM "posthog_dashboard"
  WHERE NOT (("posthog_dashboard"."deprecated_tags" IS NULL
              OR ("posthog_dashboard"."deprecated_tags" = ARRAY[]::varchar(32)[]
                  AND "posthog_dashboard"."deprecated_tags" IS NOT NULL)))

 > 78
```

```
  SELECT COUNT(*) AS "__count"
  FROM "ee_enterpriseeventdefinition"
  INNER JOIN "posthog_eventdefinition" ON ("ee_enterpriseeventdefinition"."eventdefinition_ptr_id" = "posthog_eventdefinition"."id")
  WHERE NOT (("ee_enterpriseeventdefinition"."deprecated_tags" IS NULL
              OR ("ee_enterpriseeventdefinition"."deprecated_tags" = ARRAY[]::varchar(32)[]
                  AND "ee_enterpriseeventdefinition"."deprecated_tags" IS NOT NULL)))

 > 10
```

```
  SELECT COUNT(*) AS "__count"
  FROM "ee_enterprisepropertydefinition"
  INNER JOIN "posthog_propertydefinition" ON ("ee_enterprisepropertydefinition"."propertydefinition_ptr_id" = "posthog_propertydefinition"."id")
  WHERE NOT (("ee_enterprisepropertydefinition"."deprecated_tags" IS NULL
              OR ("ee_enterprisepropertydefinition"."deprecated_tags" = ARRAY[]::varchar(32)[]
                  AND "ee_enterprisepropertydefinition"."deprecated_tags" IS NOT NULL)))

 > 0
```

Since the migration failed on `dashboard_item` and `dashboard`, I pulled in the 113 (dashboard items) + 78 (dashboards) rows from prod into python and run the for loops locally as a sanity check. [Code and results](https://gist.github.com/alexkim205/49a69746369db7891adfdc6931ff09fc) show that all tags are parsed correctly.

**Comment by alexkim205** [comment]
Also worth mentioning that the large number of one-off queries in the snapshot represent the absolute worst case scenario where 1001 objects have overlapping tags. A quick query in metabase tells me that's not going to be a problem in practice.

**Review by hazzadous — APPROVED** [review]
Looks good

🤞 the tests cover all the cases we need. It's a shame about the duplication, we seem to be a prisoner of django app structure which I think we could probably ignore to some extent (i.e. I think we could actually put this into one data migration in either ee or posthog).

Lots of yak shaving to do, happy to get in
